packages/teams-js/src/private/associatedApps.ts

```diff
@@ -0,0 +1,237 @@
+import { sendMessageToParentAsync } from '../internal/communication';
+import { ensureInitialized } from '../internal/internalAPIs';
+import { ErrorCode, SdkError, TabInstance } from '../public';
+import { runtime } from '../public/runtime';
+
+// Open questions
+// 1. I've re-used `TabInstance` from the public API, does that contain all of the information you and app developers might need?
+// 2. I didn't see any reason to add a `getTabs` function because `pages.tabs.getTabInstances`. Any reason that won't work for you?
+// 3. I've added an `AppTypes[]` param to `addAndConfigureApp` to allow for the host to show different app types to the user. Very open to changes.
+// 4. I've added empty, private `validate` functions for the threadId and TabInstance. Any validation that is possible will help prevent against
```

Comment:
This might be difficult to do on the TeamsJS layer. On the TMP side we should obviously do it.

Reply:
Yup, I'm not making it a hard requirement but something to please keep in mind: not having validation in TeamsJS for input as the source of a *large* amount of bugs and pain as we tried to expand this to other hosts. In many cases, the TeamsJS documentation and validation was missing because it was baked into the Teams codebase, which made it very difficult in practice for other hosts that wanted to support functions to figure out what they needed to do, what valid input was, etc.

Validating it in the TMP layer is better than nothing obviously, but validation in TeamsJS can help us protect against different definitions of "legal input" on other hosts, which can be a real challenge for making cross platform applications.

Like I said: it's possible there's no way to do validation in TeamsJS (although at the very least I'd like you to check for things like empty string), but the more validation you can do in this layer (or in the host sdk) the better!